Shared/PrivilegedRequest.swift

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+
+private let REQUEST_KEY_PRIVILEGED = "privileged"
+
+/**
+ Request that is allowed to load local resources.
+
+ Pages running on the local server have same origin access all resources
+ on the server, so we need to prevent arbitrary web pages from accessing
+ these resources. We do so by explicitly requiring "privileged" requests
+ in our navigation policy when loading local resources.
+
+ Be careful: creating a privileged request for an arbitrary URL provided
+ by the page will break this model. Only use a privileged request when
+ needed, and when you are sure the URL is from a trustworthy source!
+ **/
+public class PrivilegedRequest: NSMutableURLRequest {
+    public static let key = "brave_prv"
+    public static let token = UUID().uuidString.replacingOccurrences(of: "-", with: "")
+
+    override init(url URL: URL, cachePolicy: NSURLRequest.CachePolicy, timeoutInterval: TimeInterval) {
+        let modifyURL = { (url: URL) -> URL in
+            if PrivilegedRequest.isWebServerRequest(url: url),
+               let url = PrivilegedRequest.store(url: url) {
+                return url
+            }
+            return url
+        }
+
+        super.init(url: modifyURL(URL), cachePolicy: cachePolicy, timeoutInterval: timeoutInterval)
+        setPrivileged()
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        setPrivileged()
+    }
+
+    private func setPrivileged() {
+        URLProtocol.setProperty(true, forKey: REQUEST_KEY_PRIVILEGED, in: self)
+    }
+
+    private static func store(url: URL) -> URL? {
+        guard var components = URLComponents(string: url.absoluteString) else { return nil }
+
+        var queryItems = components.queryItems ?? []
+        if var item = queryItems.find({ $0.name == PrivilegedRequest.key }) {
+            item.value = PrivilegedRequest.token
+        } else {
+            let queryItem = URLQueryItem(name: PrivilegedRequest.key,
+                                         value: PrivilegedRequest.token)
+            queryItems.append(queryItem)
+        }
+
+        components.queryItems = queryItems
+        return components.url
+    }
+
+    public static func removePrivileges(url: URL) -> URL? {
+        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false),
```

Comment:
I understand the reason for doing this, but there is still a possibility of leaking the `token` in logs. I don't think it happens in this PR, but URLs are often logged for debugging purposes.

What do you think about adding the key as a headers instead?

Reply:
I would, but what would we do when we call `history.go(currentPage);` in `SessionRestore.html` since I can't add headers there :(